### PR TITLE
Replace lazer scoring with "ScoreV2"

### PR DESCRIPTION
**Pull request by smoogipoo**
One of the issues with scoring that's being consistently brought up is its linear combo scaling nature. This has triggered discussions like https://github.com/ppy/osu/discussions/20717, which has been turned into an implementation [here](https://github.com/Zyfarok/osu/tree/new-scoring).

The problem with the implementation above is that it adds a lot of complexity in an effort to make score re-calculatable, but it's not going to be unless all judgements or some metadata for them are to be stored. On top of that, it changes scoring completely unlike anything we've seen thus far, which is neither a bad thing nor a good thing.

I believe the best path forward is to return to something that everyone's familiar with that has been tried and tested. ScoreV2 seems appropriate.

For the most part, this is a 1:1 implementation of osu!stable's ScoreV2. osu!catch has not been exactly implemented, since its implementation drastically increases complexity and I've been [told](https://discord.com/channels/188630481301012481/542532504478023710/1109157966814400644) the simple combo/accuracy portions are fine, from those that have attempted to improve on the osu!stable algorithm for tournament purposes.

Of note:
- This scoring (which I'll refer to as "osu!lazer score" and _not ScoreV2_) is intended for _solo play only_. This algorithm doesn't need to be suitable for tournament play, and tournament play may have its own completely different scoring algorithm in the future.
- osu!lazer scoring should go through a period of community review after which it will be made final and remain as is.
- It's not recalculatable.

### CODE REVIEW

**Comment by bdach** [comment]
Tests need some further massaging here.

File: osu.Game/Online/Spectator/FrameHeader.cs
Comment (peppy) [inline osu.Game/Online/Spectator/FrameHeader.cs:21]: As long as we're happy with this being a breaking change (I guess it's fine).

File: osu.Game/Online/Spectator/FrameHeader.cs
Comment (peppy) [inline osu.Game/Online/Spectator/FrameHeader.cs:64]: I think I'd feel more comfortable if the `ScoreProcessorStatistics` were passed here, rather than the processor.

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs
Comment (peppy) [inline osu.Game/Rulesets/Scoring/ScoreProcessor.cs:498]: This is quite a bit of extra data to be sending in every score frame. I think it's going to probably double the bandwidth requirement of spectator (it was already pretty high so maybe a concern?)

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs
Comment (peppy) [inline osu.Game/Rulesets/Scoring/ScoreProcessor.cs:507]: This kind of variable needs xmldoc. Also maybe needs to be renamed, once I understand what it's meant to be (I think `AccuracyJudgementCount` might be more correct?)

This change removed xmldoc on other properties in this class, see above in the diff.

File: osu.Game/Scoring/Legacy/ScoreInfoExtensions.cs
Comment (peppy) [inline osu.Game/Scoring/Legacy/ScoreInfoExtensions.cs:48]: I'd like to see this split out into its own method with xmldoc explaining what the multipliers are used for and (very roughly) how they were obtained.

Right now there's no documentation explaining this.

File: osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs
Comment (peppy) [inline osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs:76]: I'm not sure I understand this. In this case, is there any reason to provide a `TotalScore.Value` update here? Internally, this triggers a `ValueChanged` which causes `updateScore` to get called, but this feels super haphazard. Like I'd hope that setting a new `GetDisplayScore` function did this instead.

And maybe the flow should be changed so the result of `GetDisplayScore` is stored to `TotalScore`, or something similar.

File: osu.Game/Screens/Play/HUD/GameplayScoreCounter.cs
Comment (peppy) [inline osu.Game/Screens/Play/HUD/GameplayScoreCounter.cs:48]: This feels a bit weird (using one bindable's `BindValueChanged` but then retrieving a completely different method).

Can you remind me what the reasoning was behind removing `Mode` from `ScoreProcessor`? I think it's one of the areas we discussed but I can't recall why this change was required.

**Review by peppy — COMMENTED** [review]
Initial pass on structure, haven't gone through the actual score V2 algorithms yet.

**Comment by Zyfarok** [comment]
For the decision to accept to go with something that is not "recalculatable", I think this is a good decision if you indeed do not want to store all combos and their associated judgment data. I'm glad your going in that direction.
And scorev2 is indeed better than lazer, so this is an improvement.

However, I hope that the "familiarity" argument will not be used against suggestions of future changes/improvements to the formula.

And more importantly, there's two things which I think are mistaken:

> This is intended for solo play only. This algorithm doesn't need to be suitable for tournament play.

I think this is a huge mistake. A formula that is best for tournaments is also best for solo play (the reverse might not be true), and it's would be simpler to have the same formula everywhere for most things in the future. The use of custom formulas should only be needed for tournaments that want to have original rule-sets (e.g., combo-only tournaments). 

> The problem with the implementation above is that it adds a lot of complexity in an effort to make score re-calculatable, but it's not going to be unless all judgements or some metadata for them are to be stored. On top of that, it changes scoring completely unlike anything we've seen thus far, which is neither a bad thing nor a good thing.

It doesn't make sense to say that when the scoring implemented in the fork is not more complex than score v2.
The fork is based on Scorev2 with :
1- One simplification to remove the impact of the position of 100/50s (in v2 if you 100/50 at the start of a combo, it has less effect than when you do it at the end, which is unjustified)
2- One tiny increase of complexity to add an exponent (~0.5) on combo (which is highly justified from the common criticism made towards v2 that 1 missing a map is still too impactful on the score) and add a acc exponent on the combo portion.
3- And finally a slight adjustment the values of v2's acc exponent.

So overall, compared to v2, saying that it "adds a lot of complexity" seems totally exaggerated. The only real "complexity" that I suggested to introduce was to be able to be able to test and adjust the combo exponent based on user feedback without the need to reset leaderboards each time, but this is not something needed permanently and can even be fully discarded if we don't want to go that way. We tried a combo exponent of 0.5, it seems to work well and we could just as well decide to use this value directly and store a single "ComboScore" just as needed for Scorev2.

**Comment by peppy** [comment]
> However, I hope that the "familiarity" argument will not be used against suggestions of future changes/improvements to the formula.

Any changes to be made will be made before things go live. We will never make changes again in the future to scoring. If you have any suggestions, please make them after this is merged within the next few months.

> I think this is a huge mistake. A formula that is best for tournaments is also best for solo play (the reverse might not be true), and it would be simpler to have the same formula everywhere for most things in the future. The use of custom formulas should only be needed for tournaments that want to have original rule-sets (e.g., combo-only tournaments).

You may be reading into this wrong. This is a disclaimer that if changes are required to scoring for tournament play, they CAN be done local to the multiplayer system. All this is saying is any changes will NOT be applied to solo play. This is non-negotiable from our end.

I ask that you wait for this to be merged and deployed rather than making this PR a discussion thread, but also ask that you follow up with feedback once we are live. There's still room to make changes and the idea here is to get the new scoring out so people can try it more easily.

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs
Comment (smoogipoo) [reply osu.Game/Rulesets/Scoring/ScoreProcessor.cs:498]: It's every 100 frames/when they get purged. Do you have a solution in mind?

Comment (peppy) [reply osu.Game/Rulesets/Scoring/ScoreProcessor.cs:498]: I think I got confused by the class name `FrameHeader` and https://github.com/ppy/osu/blob/f8830c6850128456266c82de83273204f8b74ac0/osu.Game/Rulesets/Replays/ReplayFrame.cs#L18-L23

Maybe it should be renamed to `FrameBundleHeader` or something...

File: osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs
Comment (smoogipoo) [reply osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs:76]: I think you need to try out your proposal itself, because it's not easy to make work.

File: osu.Game/Screens/Play/HUD/GameplayScoreCounter.cs
Comment (smoogipoo) [reply osu.Game/Screens/Play/HUD/GameplayScoreCounter.cs:48]: Because `ScoreProcessor` should never expose a non-standardised total score. Classic score is purely for display purposes.

**Comment by smoogipoo** [comment]
> It doesn't make sense to say that when the scoring implemented in the fork is not more complex than score v2.

You misunderstood my intentions. The complexity is in trying to make it re-calculatable. The addition of members to `JudgementResult`, `ScoreInfo` and `SoloScoreInfo`, while not simplifying any of the code of `ScoreProcessor` like `extractScoringValues()` and the various `ComputeScore()` methods that only exist for the purpose of recalculating scores.

As far as I can tell, you can't recalculate `ComboScore` without resetting leaderboards, unless we stored the judgements at the combo break points, correct? If so, that's what I meant.

It isn't helped by the implementation being different from anything _I've_ seen thus far - it makes it harder for me to go into code and refactor it, as I have to both understand and refactor the code at the same time.

Don't take it as dismissal of your code, I'm just taking a different path of returning to something that's very simple and well known (as a player but also as a developer - both peppy and I have seen this code before), and building it out from there.

Maybe the changes that you mentioned are beneficial to have and can be applied over the existing ScoreV2 implementation. That's part of the "building it out from there" point.

File: osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs
Comment (peppy) [reply osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs:76]: Figured as much. I'll keep that in mind when reviewing further.

**Comment by Zyfarok** [comment]
Thank you for your answers, all good then. Just wanted to be sure that we agreed on those points and that the plan is not to use v2 as the final lazer solo scoring. 👍  

I'll take my suggestions of improvements over score v2 into a new discussion thread once this has been merged.

File: osu.Game/Online/Spectator/FrameHeader.cs
Comment (smoogipoo) [reply osu.Game/Online/Spectator/FrameHeader.cs:21]: As a note, I'm 95% sure that _any_ change to a MessagePack model is a guaranteed breaking change, and that adding members to the end doesn't prevent it.

Comment (peppy) [reply osu.Game/Online/Spectator/FrameHeader.cs:21]: Yeah I think you're right.

File: osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs
Comment (peppy) [reply osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs:76]: I've at least fixed the part where if these two lines were reversed it would fail to update the initial display score. Probably fine for now.

**Comment by peppy** [comment]
I've done an initial pass on this and think it should be okay to move forward with.

@smoogipoo can you take a look over my changes and make sure nothing stands out as wrong?

**Review by peppy — APPROVED** [review]

**Comment by smoogipoo** [comment]
The changes are good.

**Comment by LianSheng197** [comment]
Please forgive my ignorance if I am commenting in the wrong place.

***

For the continuous update of **osu!lazer**, I think any changes are worth looking forward to, including this score calculation formula.
However, I noticed that the changelog reads:

> It isn't intending to track any past or future changes to ScoreV2
> It doesn't support recalculation.

This means that my past play scores on **osu!lazer** will continue to be preserved.

I'm a new player who only became active in April this year, because I use the `GNU/Linux (Ubuntu 22.04)` operating system and I don't use wine, so I have only played **osu!lazer** so far.
I had left grades on some pieces before the score calculation was updated. Because it is a beginner period, the highest score may be 600k ~ 800k, which usually means some misses before the update.
And now after changing the score formula, I think I may not be able to surpass the previous record in the short term, even if my accurate and combo are better than the past results.

In a better situation, I may only need to work hard for a few more months; in a worse situation, I may not be able to break through again in this life: I'm not a professional gamer, and it is impossible to become a professional gamer.

The current situation caught me off guard.
Although I understand that the focus is still on accurate and combo, but the current **osu!lazer** interface does not support sorting based on these two variables.
In the past, I liked to watch myself gradually break through yesterday's self as I practiced every day, so I attached great importance to the point system. However this score calculation update spoils that.

I have thought about clearing the local score, or switching back to version `2023.513.0` to play, maybe this will make me feel better. But as soon as I see the web profile, I feel that all this is just deceiving myself after all.

Right now, I'm confused, I don't know if I can keep playing osu! and continue to hone my skills in osu!.
Games are supposed to be enjoyable, but **osu!lazer** just isn't doing it for me right now.

I don't know if this request is reasonable, but if so, I hope **osu!lazer** will come up with some solutions in the future to solve the dilemma similar to the one I'm currently facing.

**Comment by bdach** [comment]
@LianSheng197 I understand your frustration, but please read https://github.com/ppy/osu/issues/23782#issuecomment-1583991833. Online leaderboards will get wiped to remove all scores set with the old score algorithm.

**Comment by LianSheng197** [comment]
@bdach Thank you for such a quick response.
If the online leaderboard does clear, then I'll be able to start over.

Thank you very much!

**Comment by bdach** [comment]
Do note that there is no estimate as to _when_ this reset will happen, but it _is_ planned. See https://github.com/ppy/osu/discussions/23761#discussioncomment-6096709.

**Comment by LianSheng197** [comment]
Thank you, I'm very much looking forward to that day.

**Pull request merged**
